src/zenml/services/service.py

```diff
@@ -0,0 +1,358 @@
+
+        svc = cast("BaseService", super().__call__(*args, **kwargs))
+        ServiceRegistry().register_service(svc)
+        return cast("BaseServiceMeta", svc)
+
+
+class BaseService(BaseTypedModel, metaclass=BaseServiceMeta):
+    """Base service class
+
+    This class implements generic functionality concerning the life-cycle
+    management and tracking of an external service (e.g. process, container,
+    kubernetes deployment etc.).
+
+    Attributes:
+        SERVICE_TYPE: a service type descriptor with information describing
+            the service class. Every concrete service class must define this.
+        admin_state: the administrative state of the service.
+        uuid: unique UUID identifier for the service instance.
+        config: service configuration
+        status: service status
+        endpoint: optional service endpoint
+    """
+
+    SERVICE_TYPE: ClassVar[ServiceType]
+
+    uuid: UUID = Field(default_factory=uuid4, allow_mutation=False)
+    admin_state: ServiceState = ServiceState.INACTIVE
+    config: ServiceConfig = Field(default_factory=ServiceConfig)
+    status: ServiceStatus = Field(default_factory=ServiceStatus)
+    # TODO [MEDIUM] allow multiple endpoints per service
+    endpoint: Optional[BaseServiceEndpoint]
+
+    def __init__(
+        self,
+        **attrs: Any,
+    ) -> None:
+        super().__init__(**attrs)
+        self.config.name = self.config.name or self.__class__.__name__
+
+    @abstractmethod
+    def check_status(self) -> Tuple[ServiceState, str]:
+        """Check the the current operational state of the external service.
+
+        This method should be overridden by subclasses that implement
+        concrete service tracking functionality.
+
+        Returns:
+            The operational state of the external service and a message
+            providing additional information about that state (e.g. a
+            description of the error, if one is encountered while checking the
+            service status).
+        """
+
+    def update_status(self) -> None:
+        """Check the the current operational state of the external service
+        and update the local operational status information to reflect it.
+
+        This method should be overridden by subclasses that implement
+        concrete service status tracking functionality.
+        """
+        logger.debug(
+            "Running status check for service '%s' ...",
+            self,
+        )
+        state, err = self.check_status()
+        logger.debug(
+            "Status check results for service '%s': %s [%s]",
+            self,
+            state.name,
+            err,
+        )
+        self.status.update_state(state, err)
+
+        # don't bother checking the endpoint state if the service is not active
+        if self.status.state == ServiceState.INACTIVE:
+            return
+
+        if self.endpoint:
+            self.endpoint.update_status()
+
+    def to_dict(self) -> Dict[str, Any]:
```

Comment:
I couldn't really use pydantic's `dict()` to also make `json()` work automatically, because `json()` doesn't call `dict()` under the hood, but I did manage to remove all 4 methods from the BaseService class and replace them by adding a `service_type` immutable pydantic attribute to the class automatically with the meta-class.